cjdns-core/src/encoding.rs

```diff
@@ -31,393 +32,614 @@
+
+            for i in 1..30 {
+                let prefix_len = i;
+                let bit_count = i;
+                pack.push(encoding_form(bit_count, prefix_len, prefix));
+                prefix = prefix << 1;
+            }
+
+            assert!(validate(&pack).is_ok());
+            let serialized = serialize_forms(&pack).expect("failed to serialize");
+            let deserialized = deserialize_forms(&serialized).expect("failed to deserialize");
+            assert_eq!(deserialized, pack);
         }
-        used_prefixes.insert(form.prefix);
     }
-    Ok(())
 }
 
-/// Store encoding scheme (array of `EncodingSchemeForm`) into a byte vector array (bits sequence).
-///
-/// Accepts vector of `EncodingSchemeForm`s, encodes them as bits sequence
-/// and returns the result as bytes vector.
-pub fn serialize_forms(encforms: &[EncodingSchemeForm]) -> Result<Vec<u8>> {
-    let mut result_vec: Vec<u8> = [].to_vec();
-    let mut pos = 0_u32;
-    let mut cur_byte_num = 0;
-    let mut cur_bit_num = 0_u8;
-
-    for encform in encforms {
-        // any form can be packed in u64
-        let mut acc = 0_u64;
-
-        if encform.prefix_len > 31 {
-            return Err(EncodingSchemeError::BadEncodingForm);
-        }
+mod encoding_scheme {
+    //! Routing label encoding scheme.
+
+    use std::ops::Deref;
+    use std::collections::HashSet;
+
+    use crate::encoding::errors::{SchemeValidationError, FormValidationError};
+
+    /// Encoding scheme - an iterable list of scheme forms.
+    ///
+    /// Schemes are comparable for equality, immutable, opaque and iterable.
+    #[derive(Debug, Clone, PartialEq, Eq)]
+    pub struct EncodingScheme(Vec<EncodingSchemeForm>);
+
+    /// A form of an encoding scheme. Form is used as follows to encode a director:
+    ///
+    /// ```text
+    /// [     director     ] [     form.prefix     ]
+    /// ^^^^^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^^^^^^^^
+    /// form.bit_count bits   form.prefix_len bits
+    /// ```
+    #[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+    pub struct EncodingSchemeForm {
+        bit_count: u8, // bit_count going first is important for EncodingScheme ordering
+        prefix_len: u8,
+        prefix: u32,
+    }
 
-        if encform.bit_count < 1 || encform.bit_count > 31 {
-            return Err(EncodingSchemeError::BadEncodingForm);
+    impl EncodingSchemeForm {
+        /// Instantiates `EncodingSchemeForm`.
+        ///
+        /// Returns an error in several situations:
+        /// * encoding `bit_count` value is out of valid range, which is 1..32
+        /// * `prefix_len` is too small for the provided `prefix`
+        pub fn try_new(bit_count: u8, prefix_len: u8, prefix: u32) -> Result<Self, FormValidationError> {
+            if bit_count == 0 || bit_count > 31 {
+                return Err(FormValidationError::BadBitCount);
+            }
+            if 2u32.pow(prefix_len as u32) <= 32 - prefix.leading_zeros() {
```

Comment:
You kidding me.
`(1 << prefix_len) - 1` is the maximum value for the prefix
`prefix > (1 << prefix_len) - 1` is the error condition
`prefix <= (1 << prefix_len) - 1` is the ok condition

Reply:
```
let prefix_max_value = (1 << prefix_len) - 1;
if prefix > prefix_max_value {
    return Error(....
```
for maximum readability.